src/galaxy/Economy.cpp

```diff
@@ -2,108 +2,370 @@
-		{ Lang::INDUSTRIAL_MACHINERY,
-			{ Commodity::METAL_ALLOYS, Commodity::ROBOTS },
-			ECON_INDUSTRY },
-		{ Lang::FARM_MACHINERY,
-			{ Commodity::METAL_ALLOYS, Commodity::ROBOTS },
-			ECON_INDUSTRY },
-		{ Lang::MINING_MACHINERY,
-			{ Commodity::METAL_ALLOYS, Commodity::ROBOTS },
-			ECON_INDUSTRY },
-		{ Lang::AIR_PROCESSORS,
-			{ Commodity::PLASTICS, Commodity::INDUSTRIAL_MACHINERY },
-			ECON_INDUSTRY },
-		{ Lang::SLAVES,
-			{},
-			ECON_AGRICULTURE },
-		{ Lang::HAND_WEAPONS,
-			{ Commodity::COMPUTERS },
-			ECON_INDUSTRY },
-		{ Lang::BATTLE_WEAPONS,
-			{ Commodity::INDUSTRIAL_MACHINERY, Commodity::METAL_ALLOYS },
-			ECON_INDUSTRY },
-		{ Lang::NERVE_GAS,
-			{},
-			ECON_INDUSTRY },
-		{ Lang::NARCOTICS,
-			{},
-			ECON_INDUSTRY },
-		{ Lang::MILITARY_FUEL,
-			{ Commodity::HYDROGEN },
-			ECON_INDUSTRY },
-		{ Lang::RUBBISH,
-			{},
-			ECON_INDUSTRY },
-		{ Lang::RADIOACTIVES,
-			{},
-			ECON_INDUSTRY }
+	std::vector<CommodityInfo> m_commodities;
+	std::vector<EconomyInfo> m_economies;
+
+	/* clang-format off */
+	// A dummy CommodityInfo structure for use in debugging bad EconomyIds and other foolishness.
+	CommodityInfo null_commodity = {
+		InvalidCommodityId,
+		InvalidEconomyId,
+		"NULL_COMMODITY",
+		"NULL_COMMODITY",
+		{},
+		0.0
+	};
+
+	const char *null_economy_name = "NULL_ECONOMY";
+
+	// A dummy EconomyInfo structure for use in debugging bad EconomyIds and other foolishness.
+	EconomyInfo null_economy = {
+		InvalidEconomyId,
+		null_economy_name,
+		{
+			null_economy_name,
+			null_economy_name,
+			null_economy_name,
+			null_economy_name
+		},
+		{},
+		{}
 	};
+	/* clang-format on */
+
+	std::map<CommodityId, ConsumableInfo> m_consumables;
+
+	std::map<CommodityId, IllegalInfo> m_illegalCommodities;
+
+	// map commodity names to commodity ids for loading/saving
+	// references to set/map contents are never invalidated except when pointing to deleted elements
+	std::map<std::string, CommodityId> m_commodityNameLookup;
+
+	// map economy names to economy ids for loading/saving
+	std::map<std::string, EconomyId> m_economyNameLookup;
+
+	// store the actual string data pointed to by economy / commodity names and lang keys
+	std::set<std::string> m_string_data;
+
+	const char *get_lang_key(const Json &j)
+	{
+		std::string key = j;
+		return m_string_data.emplace(key).first->c_str();
+	}
+
+	void from_json(const Json &j, CommodityInfo &out)
+	{
+		// id and name are set by LoadEconomyData
+
+		out.l10n_key = get_lang_key(j["l10n_key"]);
+
+		// this depends on m_commodityNameLookup being fully populated by looping over all commodity names
+		// and pre-assigning CommodityIds for them.
+		const Json &inputs = j["inputs"];
+		if (inputs.is_array() && inputs.size() > 0) {
+			for (size_t idx = 0; idx < inputs.size() && idx < CommodityInfo::MAX_ECON_INPUTS; idx++) {
+				const Json &input = inputs[idx];
+
+				std::string name;
+				fixed amount = fixed(1, 1);
+
+				if (input.is_array()) {
+					name = input[0];
```

Comment:
This and line 89 fail to compile in MSVC:
> E:\Games\pioneer-dev\the-eg\pioneer\src\galaxy\Economy.cpp(86): error C2593: 'operator =' is ambiguous

Reply:
`name = input[0].get<std::string>()` and `name = input.get<std::string>()` instead seems to work, but I'm not familiar with this JSON library or what's going on here.

Reply:
Yup, that's my bad. It's a limitation of the json library (and C++ implicit conversion operator overloading in general) - I'll fix it in the cleanup.